vendor/bat-native-ledger/src/bat/ledger/internal/bat_contribution.cc

```diff
@@ -92,6 +94,104 @@ std::string BatContribution::GetAnonizeProof(
   return proof;
 }
 
+void BatContribution::HasSufficientBalance(
+    ledger::HasSufficientBalanceToReconcileCallback callback) {
+  ledger_->FetchWalletProperties(
+      std::bind(&BatContribution::OnSufficientBalanceWallet,
+        this, _1, _2, callback));
+}
+
+void BatContribution::GetVerifiedAutoAmount(
+    const ledger::PublisherInfoList& publisher_list,
+    uint32_t record,
+    double balance,
+    ledger::HasSufficientBalanceToReconcileCallback callback) {
+  double ac_amount = ledger_->GetContributionAmount();
+  double total_reconcile_amount(GetAmountFromVerifiedAuto(
+      publisher_list, ac_amount));
+  if (balance < total_reconcile_amount && !publisher_list.empty()) {
+    callback(false);
+    return;
+  }
+  ledger_->GetRecurringTips(
+      std::bind(&BatContribution::GetVerifiedRecurringAmount,
+                this,
+                _1,
+                _2,
+                balance,
+                total_reconcile_amount,
+                publisher_list.empty(),
+                callback));
+}
+
+// static
+double BatContribution::GetAmountFromVerifiedAuto(
+    const ledger::PublisherInfoList& publisher_list,
+    double ac_amount) {
+  double non_verified_bat = 0.0;
+  for (const auto& publisher : publisher_list) {
+    if (!publisher->verified) {
+      non_verified_bat += (publisher->percent / 100.0) * ac_amount;
+    }
+  }
+  return ac_amount - non_verified_bat;
+}
+
+void BatContribution::GetVerifiedRecurringAmount(
+    const ledger::PublisherInfoList& publisher_list,
+    uint32_t record,
+    double balance,
+    double total_reconcile_amount,
+    bool is_ac_list_empty,
+    ledger::HasSufficientBalanceToReconcileCallback callback) {
+  if (publisher_list.empty() && is_ac_list_empty) {
```

Comment:
`is_ac_list_empty` is still needed? We now exit in AC if over, so if `publisher_list` is empty we know that we have enough funds

Reply:
Yes, because `publisher_list` here is now list from RecurringTips. `is_ac_list_empty` is value from `publisher_list` from ac table. We may still not have enough funds for recurring if ac table is empty